src/optimagic/constraints.py

```diff
@@ -0,0 +1,246 @@
+from abc import ABC, abstractmethod
+from dataclasses import KW_ONLY, dataclass
+from typing import Any, Callable
+
+import numpy as np
+import pandas as pd
+from numpy.typing import ArrayLike
+
+from optimagic.exceptions import InvalidConstraintError
+from optimagic.typing import PyTree
+
+
+class Constraint(ABC):
+    """Base class for all constraints used for subtyping."""
+
+    selector: Callable[[PyTree], PyTree]
+
+    @abstractmethod
+    def _to_dict(self) -> dict[str, Any]:
+        pass
+
+    def __post_init__(self) -> None:
+        if not callable(self.selector):
+            raise InvalidConstraintError("'selector' must be callable.")
+
+
+def identity_selector(x: PyTree) -> PyTree:
+    return x
+
+
+@dataclass(frozen=True)
+class FixedConstraint(Constraint):
```

Comment:
Can you write docstrings for each constraint? Should be faster now than in a month.